### PR TITLE
chore: add redelegate fee

### PR DESCRIPTION
**Pull request by Pythonberg1997**

### CODE REVIEW

File: contracts/BC_fusion/StakeHub.sol [hunk 1]
```diff
@@ -453,14 +457,20 @@ contract StakeHub is System, Initializable {
         _isRedelegating = 1;
         uint256 bnbAmount = IStakeCredit(srcValInfo.creditContract).unbond(delegator, shares);
         if (bnbAmount < minDelegationBNBChange) revert DelegationAmountTooSmall();
+        // check if the srcValidator has enough self delegation
+        if (IStakeCredit(srcValInfo.creditContract).getPooledBNB(srcValidator) < minSelfDelegationBNB) {
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:461]: only do such check when `srcValidator == msg.sender`

File: contracts/BC_fusion/StakeHub.sol [hunk 2]
```diff
@@ -644,6 +654,13 @@ contract StakeHub is System, Initializable {
             uint256 newUnbondPeriod = value.bytesToUint256(32);
             if (newUnbondPeriod < 3 days) revert InvalidValue(key, value);
             unbondPeriod = newUnbondPeriod;
+        } else if (key.compareStrings("redelegateFeeRate")) {
+            if (value.length != 32) revert InvalidValue(key, value);
+            uint256 newRedelegateFeeRate = value.bytesToUint256(32);
+            if (newRedelegateFeeRate > REDELEGATE_FEE_RATE_BASE || newRedelegateFeeRate == 0) {
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:660]: ==> `newRedelegateFeeRate > 100`

File: contracts/BC_fusion/StakeHub.sol [hunk 3]
```diff
@@ -453,14 +457,20 @@ contract StakeHub is System, Initializable {
         _isRedelegating = 1;
         uint256 bnbAmount = IStakeCredit(srcValInfo.creditContract).unbond(delegator, shares);
         if (bnbAmount < minDelegationBNBChange) revert DelegationAmountTooSmall();
+        // check if the srcValidator has enough self delegation
```
Comment (buddh0) [inline contracts/BC_fusion/StakeHub.sol:460]: if (delegator == srcValidator) { ？

File: contracts/BC_fusion/StakeHub.sol (hunk 2 above)
Comment (zjubfd) [reply contracts/BC_fusion/StakeHub.sol:660]: I think we can allow `newRedelegateFeeRate == 0` in future.

File: contracts/BC_fusion/StakeHub.sol (hunk 1 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:461]: Done

File: contracts/BC_fusion/StakeHub.sol (hunk 2 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol]: Done

File: contracts/BC_fusion/StakeHub.sol (hunk 3 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:460]: Done

**Review by buddh0 — APPROVED** [review]

**Review by zjubfd — APPROVED** [review]

**Pull request merged**